File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.cpp
class Solution {
private:
    int m, n;
    vector<vector<int>> direction = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

public:
    void dfs(vector<vector<int>>& heights, vector<vector<bool>>& visited, int i,
             int j) {
        visited[i][j] = true;
        for (auto& d : direction) {
            int x = i + d[0];
            int y = j + d[1];
            if (x < 0 || x >= m || y < 0 || y >= n || visited[x][y] ||
                heights[i][j] > heights[x][y])
                continue;
            dfs(heights, visited, x, y);
        }
    }
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        m = heights.size();
        n = heights[0].size();
        vector<vector<bool>> Pacific(m, vector<bool>(n, false));
        vector<vector<bool>> Atlantic(m, vector<bool>(n, false));

        for (int j = 0; j < n; j++) {
            dfs(heights, Pacific, 0, j);
            dfs(heights, Atlantic, m - 1, j);
        }
        for (int i = 0; i < m; i++) {
            dfs(heights, Pacific, i, 0);
            dfs(heights, Atlantic, i, n - 1);
        }
        vector<vector<int>> ans;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (Pacific[i][j] && Atlantic[i][j]) {
                    ans.push_back({i, j});
                }
            }
        }
        return ans;
    }
};

File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.py
class Solution:
    def __init__(self):
        self.m = 0
        self.n = 0
        self.direction = [[1, 0], [-1, 0], [0, 1], [0, -1]]


    def dfs(self, heights, visited, i, j):
        visited[i][j] = True
        for k in range(4):
            x = i + self.direction[k][0]
            y = j + self.direction[k][1]
            if (
                x < 0
                or x >= self.m
                or y < 0
                or y >= self.n
                or visited[x][y] == True
                or heights[i][j] > heights[x][y]
            ):
                continue
            self.dfs(heights, visited, x, y)

    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        self.m = len(heights)
        self.n = len(heights[0])
        Pacific = [[False for _ in range(self.n)] for _ in range(self.m)]
        Atlantic = [[False for _ in range(self.n)] for _ in range(self.m)]
        for i in range(self.m):
            self.dfs(heights, Pacific, i, 0)
            self.dfs(heights, Atlantic, i, self.n - 1)
        for j in range(self.n):
            self.dfs(heights, Pacific, 0, j)
            self.dfs(heights, Atlantic, self.m - 1, j)
        ans = []
        for i in range(self.m):
            for j in range(self.n):
                if Pacific[i][j] and Atlantic[i][j]:
                    ans.append([i, j])
        return ans
